sql/expression/auto_uuid.go

```diff
@@ -0,0 +1,106 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+
+	"github.com/dolthub/vitess/go/sqltypes"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/types"
+)
+
+// AutoUuid is an expression that captures an automatically generated UUID value and stores it in the session for
+// later retrieval. AutoUuid is intended to only be used directly on top of a UUID function.
+type AutoUuid struct {
+	UnaryExpression
+	uuidCol   *sql.Column
+	foundUuid bool
+}
+
+var _ sql.Expression = (*AutoUuid)(nil)
+var _ sql.CollationCoercible = (*AutoUuid)(nil)
+
+// NewAutoUuid creates a new AutoUuid expression. The |child| expression must be a function.UUIDFunc, but
+// because of package import cycles, we can't enforce that directly here.
+func NewAutoUuid(_ *sql.Context, col *sql.Column, child sql.Expression) *AutoUuid {
+	return &AutoUuid{
+		UnaryExpression: UnaryExpression{Child: child},
+		uuidCol:         col,
+	}
+}
+
+// IsNullable implements the Expression interface.
+func (au *AutoUuid) IsNullable() bool {
+	return false
+}
+
+// Type implements the Expression interface.
+func (au *AutoUuid) Type() sql.Type {
+	return types.MustCreateString(sqltypes.Char, 36, sql.Collation_Default)
+}
+
+// CollationCoercibility implements the interface sql.CollationCoercible.
+func (au *AutoUuid) CollationCoercibility(ctx *sql.Context) (collation sql.CollationID, coercibility byte) {
+	return sql.GetCoercibility(ctx, au.Child)
+}
+
+// Eval implements the Expression interface.
+func (au *AutoUuid) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	childResult, err := au.Child.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if !au.foundUuid {
+		uuidValue, ok := childResult.(string)
+		if !ok {
+			// This should never happen – AutoUuid should only ever be placed directly above a UUID function,
+			// so the result from eval'ing its child should *always* be a string.
+			return nil, fmt.Errorf("unexpected type for UUID value: %T", childResult)
+		}
+
+		// TODO: Setting this here means that another call to last_insert_uuid() in the same statement could
+		//       read this value too early. We should verify this isn't how MySQL behaves, and then could fix
+		//       by setting a PENDING_LAST_INSERT_UUID value in the session query info, then moving it to
+		//       LAST_INSERT_UUID in the session query info at the end of execution.
+		ctx.Session.SetLastQueryInfoString(sql.LastInsertUuid, uuidValue)
```

Comment:
I was expecting most of the state to be handled by the session. For example, if we have a prepared statement INSERT with an autoUuid node (probably common?), would only the first execution update the LastInsertUuid? We get away with stateful nodes and expressions (see `Subquery`) with the `sql.Dispose` interface, but that's probably also an anti pattern in the longterm

Reply:
Ahh... great catch on the prepared statement case. Thank you! I'll test that and add in the `sql.Dispose` fix if I can't find another way to address it.

Reply:
I added tests for prepared statements in 38b7c52f8b1da08034fabe7f4dc39c802a9544e6. It looks like they didn't need the `sql.Dispose` implementation in order to execute correctly – I believe this is because we do a smaller re-analysis and must generate a new plan tree, so the value of `foundUuid` never leaks into other executions of the prepared statement. 

I'm happy to add the `sql.Dispose` implementation if you think it'd still be safer though.

Reply:
ok cool, the latest changes to prepareds might only saved the post-binding plan, resolveInsert is going to be after that step. Might become relevant in the future, but thanks for checking!